### DISCUSSION TITLE
502 Bad Gateway

### DISCUSSION THREAD

**Discussion by barbaralau3** [FIRST-Tech-Challenge/fmltc#284] [General]
I am new the google cloud so I followed the https://github.com/FIRST-Tech-Challenge/fmltc instructions as careful as I can. I got 502 Bad Gateway after I went to my https://console.cloud.google.com/appengine?project=YOUR-PROJECT-ID.  The ending url is https://project_id.region_id.r.appspot.com/ with 502 Bad Gateway in website.  Does anyone know what I can do to correct this mistake?
Thank you very much!

**Comment by lizlooney** [comment]
Did you get 502 when you went to the cloud console, or when you went to https://project_id.region_id.r.appspot.com/?

If you got the 502 when you went to ...appspot..., check the logs in the cloud console. You might see a stack trace there.

**Comment by barbaralau3** [comment]
Thank you for your response. I got the 502 fixed. And I am able to login
using the FTC fmltc login page. However, I am getting the following error
that I cannot fix for 2 days. I am new to Google Cloud and I appreciate
your help. I got this error right after I sign in in fmltc login page:

Thank you again for your help.

**Reply by lizlooney** [reply]
I don't see what error you got. Did you forget to put it in the message here?

**Comment by barbaralau3** [comment]
Hi, Liz:

I got the following error message:
iam.gserviceaccount.com does not have storage.buckets.get access to the
Google Cloud Storage bucket.'
<https://serverfault.com/questions/1112762/error-iam-gserviceaccount-com-does-not-have-storage-buckets-get-access-to-the-g>

Please see the attached screenshot. Thank you very much!

**Reply by barbaralau3** [reply]
![Screenshot from 2022-10-10 16-46-28](https://user-images.githubusercontent.com/6933646/195161936-4e72ddb1-9c4b-4728-ad10-16ffbb00ffee.png)

**Reply by lizlooney** [reply]
This is helpful!

So the name of the storage bucket is ftc10138powerplay-blobs. That's good.
The name of the service account is ftc10138powerplay.iam.gservice.account.com. That's good too.

Hold on a second. It will take me a few minutes to look up the URL you need to go to on the cloud console to fix this.

**Reply by lizlooney** [reply]
Go to https://console.cloud.google.com/iam-admin/iam?project=ftc10138powerplay.

In the center area there should be a heading "VIEW BY PRINCIPALS"
In the column "Principal", find ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gservice.account.com.
In that row, what is in the Role column?

**Reply by barbaralau3** [reply]
It is "Computer Storage Admin". I cannot find any "Storage Legacy Bucket.XXX" role in IAM & Admin's role selection box. I filtered "storage". Apparently "Computer Storage Admin" is one of the option. I just double checked again. I will switch to "Storage Admin" and try again. Thank you very much!

**Reply by barbaralau3** [reply]
I tried again. Now I got another error: FailedPrecondition('no matching index found.'). I retype the Deploy step1 in ReadMe: scripts/deploy_indexes.sh. I still got the FailedPrecondition('no matching index found.'). Attached are the screenshot for current updated IAM & Admin principal's role "[ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gservice.account.com](mailto:ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gservice.account.com)." and the login error screenshot.
![Screenshot from 2022-10-11 13-39-56](https://user-images.githubusercontent.com/6933646/195173648-049b0f1b-b913-4b7d-8bfd-b2067c99e619.png)
![Screenshot from 2022-10-11 13-39-41](https://user-images.githubusercontent.com/6933646/195173669-42bbfc0c-b441-4999-83ff-fc6f20d08ba3.png)

**Comment by barbaralau3** [comment]
Hi, Liz:

This morning I got a different error after I logged in. Please advise or
give some hint where I should start to fix it. Attached is the screenshot:

Thank you very much!

**Reply by barbaralau3** [reply]
![Screenshot from 2022-10-11 08-40-39](https://user-images.githubusercontent.com/6933646/195162054-79be754f-7dc0-4f48-8cdf-7e29c17d96d6.png)

**Reply by lizlooney** [reply]
It looks like you didn't create the database indexes.
That's step 1 under [Deploy everything](https://github.com/FIRST-Tech-Challenge/fmltc#deploy-everything).
Make sure you do that step.

**Comment by lizlooney** [comment]
Apparently, github does not accept email attachments. So if you want to attach screenshots, you need to go to [this discussion](https://github.com/FIRST-Tech-Challenge/fmltc/discussions/284) and drag/drop or paste them into a reply message.

**Reply by barbaralau3** [reply]
Liz:
I am updating to you about the current situation of my login to Google Cloud from fmltc.

1) I reviewed the ReadMe in https://github.com/FIRST-Tech-Challenge/fmltc. The error of  '
ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gserviceaccount.com does not have storage.buckets.get access to the Google Cloud Storage bucket.' still exists. It seems the error is related to IAM server account, and it was created in step 7 in the ReadMe. If I can view the fmltc login page, does it mean I don't have a problem with step 1-12 in ReadMe? I tried to narrow down the issue and find the cause.

2) I also checked that the following roles are associated with the "storage.buckets.get" permission:
Storage Legacy Bucket Reader
Storage Legacy Bucket Writer]
Storage Legacy Bucket Owner
Storage Admin 

So I opened the "Cloud Storage/Bucket" site and try to assign the roles to the principals. Here is my current  bucket list:
ftc10138powerplay 
ftc10138powerplay-action-parameters 
ftc10138powerplay-blobs 
[ftc10138powerplay.appspot.com](http://ftc10138powerplay.appspot.com/) 
gcf-sources-97972005072-us-central1 
[staging.ftc10138powerplay.appspot.com](http://staging.ftc10138powerplay.appspot.com/) 
[us.artifacts.ftc10138powerplay.appspot.com](http://us.artifacts.ftc10138powerplay.appspot.com/)

These buckets were created in ReadMe step #10. Please review the list as they are correctly created as instructed in ReadMe. 

I attached the screen print of the permission in the "ftc10138powerplay" bucket.
![Screenshot from 2022-10-11 12-48-36](https://user-images.githubusercontent.com/6933646/195163893-af3c3a93-a419-4f05-94ab-0c0afd7b1870.png)

3) I tried to assigned the roles of Storage Legacy Bucket Reader/Writer/Owner and storage Admin to this principal: ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gserviceaccount.com in bucket. The error of 'ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gserviceaccount.com does not have storage.buckets.get access to the Google Cloud Storage bucket.' still persists. 

4) I also went to IAM's  "Permissions for project "ftc10138powerplay"". I couldn't find any "Storage Legacy Bucket XXX" roles. I assigned the closest role that is related to "Storage" -> "Computer Storage Admin". Still no luck. Attached is the screen print of Permissions for project "ftc10138powerplay" in IAM & Admin:
 ![Screenshot from 2022-10-11 12-54-40](https://user-images.githubusercontent.com/6933646/195165414-330062e2-d293-4eeb-9f47-3a48baeb23fd.png)

Please advise and let me know how can I fix this issue. Thank you very much!

**Reply by lizlooney** [reply]
In your first screenshot, you are looking at the permissions for the wrong bucket. The bucket that is the problem is ftc10138powerplay-blobs.

In the second screenshot, I can see that the principal ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gserviceaccount.com does NOT have the Storage Admin Role.

You need to do this:
In the IAM page (second screenshot):
Click on the pencil in the row for ftc10138powerplay-serv-acct@ftc10138powerplay.iam.gserviceaccount.com 
Click "+ ADD ANOTHER ROLE"
Type Storage Admin. (if a dropdown appears, find "Storage Admin").
Click Save.

**Reply by barbaralau3** [reply]
Thank you Liz. I modified the roles. The error is still here: FailedPrecondition('no matching index found.'). Do I need to rerun the source env_setup.sh & scripts/deploy_indexes.sh before each "gcloud app browser"?

Here is the screen print for IAM & Admin:
![Screenshot from 2022-10-11 13-59-30](https://user-images.githubusercontent.com/6933646/195176648-6a8ad2a8-2e7a-4d56-a0a0-f313b47e195d.png)

Here is the screen print for ftc10138powerplay-blobs's bucket:
![Screenshot from 2022-10-11 14-00-46](https://user-images.githubusercontent.com/6933646/195176889-0e02f13e-d36e-41d0-98f1-e0638b5f0ea9.png)

**Comment by lizlooney** [comment]
OK. So I think Google Cloud Storage is good now.

The indexes are for the firestore database. Unfortunately, I don't know how to see those in the cloud console. (In my own fmltc projects, the Cloud Console page for Firestore Indexes says "No rows to display". I've never understood why.)

Instead, let's look at the logs from appengine. Here's what to do:
Go to https://ftc10138powerplay.uc.r.appspot.com (this will repeat the error so it is fresh in the logs)
Go to https://console.cloud.google.com/logs/query?project=ftc10138powerplay
In the "Query results" section, look for the word "traceback" or "no matching index"
Expand the rows that have those words and post a screenshot.

**Reply by barbaralau3** [reply]
Here is the log message:
CRITICAL:root:capture_exception traceback: Traceback (most recent call last): ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 65, in error_remapped_callable ...     return callable_(*args, **kwargs) ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/grpc/_channel.py", line 946, in __call__ ...     return _end_unary_response_blocking(state, call, False, None) ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking ...     raise _InactiveRpcError(state) ... grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with: ... 	status = StatusCode.FAILED_PRECONDITION ... 	details = "no matching index found." ... 	debug_error_string = "UNKNOWN:Error received from peer ipv4:64.233.182.95:443 {created_time:"2022-10-11T19:39:12.036000234+00:00", grpc_status:9, grpc_message:"no matching index found."}" ... > ...  ... The above exception was the direct cause of the following exception: ...  ... Traceback (most recent call last): ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request ...     rv = self.dispatch_request() ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request ...     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args) ...   File "/workspace/wrappers.py", line 85, in wrapper ...     return func(*args, **kwargs) ...   File "/workspace/wrappers.py", line 36, in wrapper ...     return func(*args, **kwargs) ...   File "/workspace/app_engine.py", line 539, in index ...     team_uuid = team_info.retrieve_team_uuid(flask.session, flask.request) ...   File "/workspace/team_info.py", line 99, in retrieve_team_uuid ...     team_uuid = storage.retrieve_team_uuid(program, team_number) ...   File "/workspace/storage.py", line 85, in retrieve_team_uuid ...     team_entities = list(query.fetch(1)) ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/page_iterator.py", line 208, in _items_iter ...     for page in self._page_iter(increment=False): ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/page_iterator.py", line 244, in _page_iter ...     page = self._next_page() ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/cloud/datastore/query.py", line 578, in _next_page ...     response_pb = self.client._datastore_api.run_query( ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/cloud/datastore_v1/services/datastore/client.py", line 450, in run_query ...     response = rpc(request, retry=retry, timeout=timeout, metadata=metadata,) ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/gapic_v1/method.py", line 154, in __call__ ...     return wrapped_func(*args, **kwargs) ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/retry.py", line 283, in retry_wrapped_func ...     return retry_target( ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/retry.py", line 190, in retry_target ...     return target() ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 67, in error_remapped_callable ...     raise exceptions.from_grpc_error(exc) from exc ... google.api_core.exceptions.FailedPrecondition: 400 no matching index found. ...

**Comment by lizlooney** [comment]
Or, instead of a screenshot, can you copy and paste the text? That way I copy the text to an editor where I can reformat it to see it better.

**Reply by barbaralau3** [reply]
Here is the log message:
CRITICAL:root:capture_exception traceback: Traceback (most recent call last): ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 65, in error_remapped_callable ... return callable_(*args, **kwargs) ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/grpc/_channel.py", line 946, in call ... return _end_unary_response_blocking(state, call, False, None) ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking ... raise _InactiveRpcError(state) ... grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with: ... status = StatusCode.FAILED_PRECONDITION ... details = "no matching index found." ... debug_error_string = "UNKNOWN:Error received from peer ipv4:64.233.182.95:443 {created_time:"2022-10-11T19:39:12.036000234+00:00", grpc_status:9, grpc_message:"no matching index found."}" ... > ... ... The above exception was the direct cause of the following exception: ... ... Traceback (most recent call last): ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request ... rv = self.dispatch_request() ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request ... return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args) ... File "/workspace/wrappers.py", line 85, in wrapper ... return func(*args, **kwargs) ... File "/workspace/wrappers.py", line 36, in wrapper ... return func(*args, **kwargs) ... File "/workspace/app_engine.py", line 539, in index ... team_uuid = team_info.retrieve_team_uuid(flask.session, flask.request) ... File "/workspace/team_info.py", line 99, in retrieve_team_uuid ... team_uuid = storage.retrieve_team_uuid(program, team_number) ... File "/workspace/storage.py", line 85, in retrieve_team_uuid ... team_entities = list(query.fetch(1)) ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/page_iterator.py", line 208, in _items_iter ... for page in self._page_iter(increment=False): ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/page_iterator.py", line 244, in _page_iter ... page = self._next_page() ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/cloud/datastore/query.py", line 578, in _next_page ... response_pb = self.client._datastore_api.run_query( ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/cloud/datastore_v1/services/datastore/client.py", line 450, in run_query ... response = rpc(request, retry=retry, timeout=timeout, metadata=metadata,) ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/gapic_v1/method.py", line 154, in call ... return wrapped_func(*args, **kwargs) ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/retry.py", line 283, in retry_wrapped_func ... return retry_target( ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/retry.py", line 190, in retry_target ... return target() ... File "/layers/google.python.pip/pip/lib/python3.9/site-packages/google/api_core/grpc_helpers.py", line 67, in error_remapped_callable ... raise exceptions.from_grpc_error(exc) from exc ... google.api_core.exceptions.FailedPrecondition: 400 no matching index found. ...

**Reply by lizlooney** [reply]
Great! That error is exactly the problem. Here's how I investigate this.
I look for the last place that says `File "/workspace/`
In this case, that's this:
`File "/workspace/storage.py", line 85, in retrieve_team_uuid ... team_entities = list(query.fetch(1))`

Then I look at [that code](https://github.com/FIRST-Tech-Challenge/fmltc/blob/main/server/app_engine/storage.py#L85) and see that it's doing a query with filters on the `program` and `team_number` fields.

Then I look in the [index file](https://github.com/FIRST-Tech-Challenge/fmltc/blob/main/index.yaml) and see that the index we are trying to use is declared on line 4.

So, that's all good.

Now, when you ran
```
source env_setup.sh
scripts/deploy_indexes.sh
```
was there any output or errors?

**Reply by barbaralau3** [reply]
I got the following for source env_setup.sh:
FMLTC_GCLOUD_PROJECT_ID is ftc10138powerplay
FMLTC_CLOSURE_COMPILER_JAR is ../closure-compiler/closure-compiler-v20200406.jar
FMLTC_CLOSURE_LIBRARY_FOLDER is ../closure-library/closure-library-20200406

I got no output for scripts/deploy_indexes.sh

**Reply by lizlooney** [reply]
OK. It looks like the deploy_indexes.sh script doesn't deploy if it thinks it already did it.

Please do
```
source env_setup.sh
gcloud -q datastore indexes create index.yaml
```

The output should be something like this:
```
Configurations to update:

descriptor:      [index.yaml]
type:            [datastore indexes]
target project:  [ftc10138powerplay]


....done.  
```

**Comment by barbaralau3** [comment]
Please Wait for 20 mins I am outside. Will do asap.

**Reply by lizlooney** [reply]
Sure!

**Reply by barbaralau3** [reply]
Liz, I got the screen:
![Screenshot from 2022-10-11 16-52-14](https://user-images.githubusercontent.com/6933646/195205173-39f853ac-ff64-4d1c-8772-5ff7a65b84a5.png)

Thank you for your help. I will do the testing tonight. Have a nice evening. Really appreciate your help, Liz!

**Comment by lizlooney** [comment]
That's fantastic!!!

I will update the readme for these two problems:

1. I'll add a command to give the Storage Admin role to the service account.
2. I'll document that deploying the indexes takes a long time and is asynchronous, so the indexes might not be ready until several hours (a day?) after running deploy_indexes.sh.

**Reply by barbaralau3** [reply]
Hi, Liz:
I started upload video to the cloud, however, I got the following errors from log file:
{
httpRequest: {8}
insertId: "6349d2810008175031ac4f00"
labels: {1}
logName: "projects/ftc10138powerplay/logs/cloudfunctions.googleapis.com%2Frequests"
receiveTimestamp: "2022-10-14T21:20:01.535166928Z"
resource: {2}
severity: "ERROR"
spanId: "572665622061577254"
textPayload: "The request was aborted because there was no available instance. Additional troubleshooting documentation can be found at: https://cloud.google.com/functions/docs/troubleshooting#scalability"
timestamp: "2022-10-14T21:20:01.530256Z"
trace: "projects/ftc10138powerplay/traces/2f6c664d7bf063567740a8f304051f3c"
traceSampled: true
}

**Reply by barbaralau3** [reply]
The further down error message: 
google.api_core.exceptions.PermissionDenied: 403 Cloud Datastore API has not been used in project before

Both of Cloud Firestore/Datastore API shown the activity around 10/14, both displayed 200 as color blue(middle of page, far right, )... I manually create database Firestore on Native mode in fmltc's ReadMe. I never create Firestore on Datastore mode. I do not know why both Cloud Firestore/Datastore API displayed and recorded activities:
![Screenshot from 2022-10-14 21-24-41](https://user-images.githubusercontent.com/6933646/195965002-e57b888b-739a-4ba1-8da8-59aaf0c67509.png)
 
![Screenshot from 2022-10-14 21-24-23](https://user-images.githubusercontent.com/6933646/195965033-d728366f-039d-427f-9ed3-cfb27fa23321.png)

**Reply by barbaralau3** [reply]
It looks like there were changes database configuration in google side after fmltc was build in last year. Firestore can be created by command line in 2022 and default to Native mode. https://cloud.google.com/sdk/gcloud/reference/firestore/databases/create. Does creating index is no longer required in Firestore...? Besides the deploy index command line, there are multiple places in current fmltc have Datastore information. For example in main.tf, 

My question is, the current fmltc on FTC site works fine. What is the difference between the current instructions on ReadMe.md and the actual configuration and implementation on the fmltc application?

I also knew someone who was able to train the datasets from fmltc last year (2021), any documented changes about connect to database in fmltc? 

Really appreciate if you have any idea how to fix this issue. Thank you very much!

**Comment by barbaralau3** [comment]
Understand now. It is great to learn from you and this processing . There
is one last thing if you might interested:

There is error after I rerun the actual command in deploy_indexes.sh,
“Failed Precondition-the index for the query is not ready to serve…”. The
was solved by assigning all principles the role of owner. I googled and it
works for me. I am not sure why… is there a document you can recommend to
me for brief details of the what we did today? It’s okay if you don’t, I
can try to google. Really appreciate your time and help today!

**Comment by lizlooney** [comment]
I think that the “Failed Precondition-the index for the query is not ready to serve…”. error just indicates that (as it says) the index is not ready. I believe that adding the role of owner to all principles didn't make it work... It's just that by the time you'd done that, the index was ready.

For documentation about what we did today:
[IAM roles for Cloud Storage](https://cloud.google.com/storage/docs/access-control/iam-roles)

I tried to find documentation about the indexes, but I can't find any documentation that makes sense. I think when I created fmltc, I wanted to use Datastore and then Google replaced that with Firestore. I think for a while I was using "Firestore in Datastore Mode", but the cloud console says I'm using "Firestore in Native Mode". I think the index.yaml file and the command that is in deploy_indexes.sh are typically used for Datastore. I'm not even sure why that command works if the database is actually Firestore in Native Mode. I'll see if I can figure that out.

**Reply by barbaralau3** [reply]
Hi, Liz:
I am new to google cloud so I am learning anything that's new to me in these couple days. Just regard to what you mentioned yesterday, there is a page might help me on understand the Firestore in Native or Datastore mode in a theoretical level: https://cloud.google.com/datastore/docs/firestore-or-datastore. I needed to learn how exactly the application connects to Firestore on cloud and many more... This cloud is a totally different way to develop an application than before...

**Comment by barbaralau3** [comment]
Thank you for the update and the document , Liz.

Just wondering, do you work with FTC or Google Cloud? Our team want to
train dataset more than team element this year. Is there a limit on how
many model we can train on google cloud?  Second question is once we get
the models (assume more than one), does the running and detecting object
using tensor flow fast enough in the human drive period?

It’s interesting to see how the index.yaml file work with database.

Thank you again. Have a nice evening,

**Comment by lizlooney** [comment]
I am on the FTC Tech Team, a group of volunteers who help the engineers at FIRST to create the [FTC SDK](https://github.com/FIRST-Tech-Challenge/FtcRobotController) and [fmltc](https://github.com/FIRST-Tech-Challenge/fmltc).

When you use your own google cloud project (ftc10138powerplay), the limit is money. You may have received a certain amount of free credits from Google when you created the project. When those run out, it will start costing real money.
In my experience, it isn't that expensive. In the project I use for developing fmltc, the cost of using the service to upload videos, label frames, create datasets, and train about 5 models was about $37. I'm not sure if my usage is typical though because I usually don't care if my models work very well. I don't usually have as many video frames as some people do.

As to whether it is fast enough during the tele-op part of the matches, that's a good question. In the FTC SDK, there's an object tracker that is used to track objects on the camera frames that come in while the tensorflow object detection is busy. That makes it pretty fast. You'll have to try it out and see.

**Comment by barbaralau3** [comment]
Awesome, thanks for the info! Again, really appreciate you time and
help with this!

**Comment by barbaralau3** [comment]
I am glad I am able to run the fmltc now after found out what cause the errors existed in past couple days. However, I got new errors when I clicked the "Start training" button. I am still not very familiar with the IAM and the error seems came from the IAM. I have more than 5 principals in IAM. I am not sure which one I need to check the permission. Just in case you have any clue about the following error message:

CRITICAL:root:capture_exception traceback: Traceback (most recent call last): ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/flask/app.py", line 1516, in full_dispatch_request ...     rv = self.dispatch_request() ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/flask/app.py", line 1502, in dispatch_request ...     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args) ...   File "/workspace/wrappers.py", line 85, in wrapper ...     return func(*args, **kwargs) ...   File "/workspace/wrappers.py", line 45, in wrapper ...     return func(*args, **kwargs) ...   File "/workspace/app_engine.py", line 1290, in start_training_model ...     model_entity = model_trainer.start_training_model(team_uuid, description, dataset_uuids, ...   File "/workspace/model_trainer.py", line 307, in start_training_model ...     eval_job_response = ml.projects().jobs().create(parent=parent, body=eval_job).execute() ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/googleapiclient/_helpers.py", line 134, in positional_wrapper ...     return wrapped(*args, **kwargs) ...   File "/layers/google.python.pip/pip/lib/python3.9/site-packages/googleapiclient/http.py", line 907, in execute ...     raise HttpError(resp, content, uri=self.uri) ... googleapiclient.errors.HttpError: <HttpError 403 when requesting https://ml.googleapis.com/v1/projects/XXXXpowerplay/jobs?alt=json returned "Access to project denied. This might be a transient error and a retry may succeed. If the error persists, please check the IAM permissions on your project."> .

Thank you very much.

**Reply by lizlooney** [reply]
I sounds like the app engine service account does not have permission to create a machine learning job.

**Reply by barbaralau3** [reply]
I was spending a while on searching on the API, permission and fmltc project for any clue... I reviewed the ftc fmlt instruction again and I was able to find something that I didn't understand when built up the whole application for deployment. It was up and running now. Thank you for your reply, Liz!

**Reply by barbaralau3** [reply]
Liz:
I have a question for tflite file. The PowerPlay.tflite model (from FtcRobotController ConceptTensorFlowObjectDetectionWebcam.java) can recognize 3 different labels(models) which are Bolt, Bulb, Panel. 
I got one model for one type of element, so if I trained 3 models, I will got three tflite files. How can I consolidate three tflite files into some file like PowerPlay.tflite that can recognize 3 patterns? I know this is a bit out of your scope and not related to Google Cloud... Just in case you have a clue. Thank you!

**Reply by cmacfarl** [reply]
> Liz: I have a question for tflite file. The PowerPlay.tflite model (from FtcRobotController ConceptTensorFlowObjectDetectionWebcam.java) can recognize 3 different labels(models) which are Bolt, Bulb, Panel. I got one model for one type of element, so if I trained 3 models, I will got three tflite files. How can I consolidate three tflite files into some file like PowerPlay.tflite that can recognize 3 patterns? I know this is a bit out of your scope and not related to Google Cloud... Just in case you have a clue. Thank you!

Barbara: As we move to more general questions about model training, and less specifically about hosting your own instance of fmltc,  it is probably more appropriate to ask them over in the community forum where we have dedicated categories for machine learning.  This question in particular is one that I think many people may have and the answer would be a useful addition to the forum's knowledge base.

The ftc community forums are here: https://ftc-community.firstinspires.org/

**Comment by texasdiaz** [comment]
You cannot combine TFLITE files. If you think about it like neural networks, which is exactly what it is, one neural network might be incredibly tuned for one thing but they need special training to not recognize other similar objects as that thing. That's why you need to create a dataset with ALL of the labels you want - during training the model will then learn to differentiate between objects. If you have videos labeled with individual labels, try creating a dataset with ALL of your videos. That's how I did it with all 3 images.

That's assuming, of course, that your videos individually don't include other images. If so, you need to label them or else you might confuse the network training. Anything not labeled is specifically marked to be ignored, and if one video frame has something ignored and another video frame has something labeled, that's going to lead to a bad model.

**Comment by barbaralau3** [comment]
Thank you for your reply and advice. That is very helpful!

**Comment by barbaralau3** [comment]
Hi, Liz: 

We tested on private fmltc instance and when the model failed, we tested on FTC's fmltc instance and it worked. The batch size is different between the two instances. Maybe when we switch from TPU to GPU in private fmltc instance, the smaller batch size/ more steps work better? 
Currently I got the following in my IAM:
1)[developer.gserviceaccount.com](http://developer.gserviceaccount.com/)
2)[cloudbuild.gserviceaccount.com](http://cloudbuild.gserviceaccount.com/)
3)[cloudservices.gserviceaccount.com](http://cloudservices.gserviceaccount.com/)
4)xxx.iam.gserviceaccount.com](http://xxx.iam.gserviceaccount.com/)
5)[appspot.gserviceaccount.com](http://appspot.gserviceaccount.com/)
6)[cloud-tpu.iam.gserviceaccount.com](http://cloud-tpu.iam.gserviceaccount.com/)

Would you please advice if we can do anything in IAM and assign the admin role to firestore? Thank you

**Reply by texasdiaz** [reply]
@lizlooney Specifically, I think @barbaralau3 is trying to edit the Firestore to be able to create a Configuration, whereas she can then set the ``use_tpu`` token to ``false`` so that her instance can switch to GPU. While I do not think it will help her specific use-case, at least she'd get better granularity from the metrics since more steps == more data points. I just don't know how to help her to gain admin access to the Firestore to make changes.
